### PR TITLE
feat:[CDS-51185]: Create Terraform Cloud run step manager side

### PR DESCRIPTION
**Pull request by alebuh**
## Describe your changes
Implementation of Create Terraform Cloud Run step (without any logic on the Manager or Delegate side)
Tech spec: https://harness.atlassian.net/wiki/spaces/CDP/pages/21186806281/TechSpec+CDS-4779+Flows+API+calls+for+Terraform+Cloud+Support+in+Harness
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43075)

### CODE REVIEW

**Comment by alebuh** [comment]
trigger ti1

**Review by tmacari-harness — CHANGES_REQUESTED** [review]
> Reviewed 40 of 43 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43075)* status: 43 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/pipeline/steptype/NGStepType.java` line 99 at r2](https://reviewable.io/reviews/harness/harness-core/43075#-NMn2sZy2tiHqxwA0-9n:-NMn2sZy2tiHqxwA0-9o:b-866nvk) ([raw file](https://github.com/harness/harness-core/blob/a7dfc5dfe0e1a8dd5ffd76648ef5c0c6884853b4/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/pipeline/steptype/NGStepType.java#L99)):*
> ```Java
>       "Infrastructure Provisioners/Shell Script Provisioner", StepSpecTypeConstants.PLACEHOLDER),
>   @JsonProperty(StepSpecTypeConstants.TERRAFORM_CLOUD_RUN)
>   TERRAFORM_CLOUD_RUN("Terraform cloud run", Arrays.asList(ServiceDefinitionType.values()),
> ```

If that's gonna be displayed somewhere, let's make it like "Terraform Cloud Run".
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDNGPlanCreatorProvider.java` line 1188 at r2](https://reviewable.io/reviews/harness/harness-core/43075#-NMn8jlL2mKSw7x0buUo:-NMn8jlL2mKSw7x0buUp:b-duh536) ([raw file](https://github.com/harness/harness-core/blob/a7dfc5dfe0e1a8dd5ffd76648ef5c0c6884853b4/125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDNGPlanCreatorProvider.java#L1188)):*
> ```Java
>             .build();
> 
>     StepInfo terraformCloudRun = StepInfo.newBuilder()
> ```

feartureRestrictionName is missed even it it's configured in 925-enforcement-service/src/main/resources/features/cd_features.yml
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudRunStepParameters.java` line 30 at r2](https://reviewable.io/reviews/harness/harness-core/43075#-NMnB4O--vwfc4GaxtjV:-NMnB4O--vwfc4GaxtjW:bycqn80) ([raw file](https://github.com/harness/harness-core/blob/a7dfc5dfe0e1a8dd5ffd76648ef5c0c6884853b4/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudRunStepParameters.java#L30)):*
> ```Java
> public class TerraformCloudRunStepParameters implements SpecParameters {
>   ParameterField<List<TaskSelectorYaml>> delegateSelectors;
>   String stepFqn;
> ```

let's get rid of stepFqn  we don't need to build any logic around it (we have a tech debt item to get rid of it in current terraform integration)

File: 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudRunSpecParameters.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
+ */
+
+package io.harness.cdng.provision.terraformcloud;
+
+import io.harness.pms.yaml.ParameterField;
+
+import java.util.List;
+
+public interface TerraformCloudRunSpecParameters {
```
Comment (vlica-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudRunSpecParameters.java:14]: @OwnedBy(HarnessTeam.CDP)

File: 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudRunStepInfo.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+@JsonTypeName(StepSpecTypeConstants.TERRAFORM_CLOUD_RUN)
+@FieldDefaults(level = AccessLevel.PRIVATE)
+@RecasterAlias("io.harness.cdng.provision.terraformcloud.TerraformCloudRunStepInfo")
+public class TerraformCloudRunStepInfo implements CDAbstractStepInfo, WithConnectorRef {
+  @JsonProperty(YamlNode.UUID_FIELD_NAME)
+  @Getter(onMethod_ = { @ApiModelProperty(hidden = true) })
+  @ApiModelProperty(hidden = true)
+  private String uuid;
+
+  @NotNull TerraformCloudRunType runType;
```
Comment (vlica-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudRunStepInfo.java:63]: Shoudn't this TerraformCloudRunType have a @JsonProperty?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/runspecs/TerraformCloudRefreshSpec.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+@FieldDefaults(level = AccessLevel.PRIVATE)
+@OwnedBy(HarnessTeam.CDP)
+@FieldNameConstants(innerTypeName = "TerraformCloudRefreshSpecKeys")
+@RecasterAlias("io.harness.cdng.provision.terraformcloud.runspecs.TerraformCloudRefreshSpec")
+public class TerraformCloudRefreshSpec extends TerraformCloudRunExecutionSpec {
+  @Getter(onMethod_ = { @ApiModelProperty(hidden = true) }) @ApiModelProperty(hidden = true) String uuid;
+  @NotNull @ApiModelProperty(dataType = SwaggerConstants.STRING_CLASSPATH) ParameterField<String> connectorRef;
+  @NotNull @ApiModelProperty(dataType = SwaggerConstants.STRING_CLASSPATH) ParameterField<String> organization;
+  @NotNull @ApiModelProperty(dataType = SwaggerConstants.STRING_CLASSPATH) ParameterField<String> workspace;
+  @ApiModelProperty(dataType = BOOLEAN_CLASSPATH) @YamlSchemaTypes({string}) ParameterField<Boolean> discardPendingRuns;
```
Comment (vlica-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/runspecs/TerraformCloudRefreshSpec.java:49]: Shouls we have @NotNull for discardPendingRuns? I saw in other classed you add this.

File: 125-cd-nextgen/src/main/java/io/harness/serializer/kryo/NGKryoRegistrar.java [hunk 4]
```diff
@@ -264,5 +265,7 @@ public void register(Kryo kryo) {
     kryo.register(K8sDryRunManifestStepInfo.class, 12644);
     kryo.register(K8sDryRunManifestStepParameters.class, 12645);
     kryo.register(K8sDryRunManifestOutcome.class, 12646);
+
+    kryo.register(TerraformCloudRunStepInfo.class, 12656);
```
Comment (vlica-harness) [inline 125-cd-nextgen/src/main/java/io/harness/serializer/kryo/NGKryoRegistrar.java:269]: please keep the order for kryo

**Review by alebuh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43075)* status: 43 file(s) to review. 1 change requests. 6. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/pipeline/steptype/NGStepType.java` line 99 at r2](https://reviewable.io/reviews/harness/harness-core/43075#-NMn2sZy2tiHqxwA0-9n:-NMoU0MUChYC7JFnOf3J:b-7rl0jq) ([raw file](https://github.com/harness/harness-core/blob/a7dfc5dfe0e1a8dd5ffd76648ef5c0c6884853b4/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/pipeline/steptype/NGStepType.java#L99)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

If that's gonna be displayed somewhere, let's make it like "Terraform Cloud Run".
</blockquote></details>

done!
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDNGPlanCreatorProvider.java` line 1188 at r2](https://reviewable.io/reviews/harness/harness-core/43075#-NMn8jlL2mKSw7x0buUo:-NMoU4C13ZsBuabl94g8:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/a7dfc5dfe0e1a8dd5ffd76648ef5c0c6884853b4/125-cd-nextgen/src/main/java/io/harness/cdng/creator/CDNGPlanCreatorProvider.java#L1188)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

feartureRestrictionName is missed even it it's configured in 925-enforcement-service/src/main/resources/features/cd_features.yml
</blockquote></details>

done
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudRunSpecParameters.java` line 14 at r2](https://reviewable.io/reviews/harness/harness-core/43075#-NMcy3UqEEtQMbLyS1D_-r2-14:-NMoU6j5Chg1F9O23lGJ:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/a7dfc5dfe0e1a8dd5ffd76648ef5c0c6884853b4/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudRunSpecParameters.java#L14)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

@OwnedBy(HarnessTeam.CDP)
</blockquote></details>

done
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudRunStepInfo.java` line 63 at r2](https://reviewable.io/reviews/harness/harness-core/43075#-NMcy3UqEEtQMbLyS1Da-r2-63:-NMoU8st2FRHu1YCFF1h:b-lm1ekn) ([raw file](https://github.com/harness/harness-core/blob/a7dfc5dfe0e1a8dd5ffd76648ef5c0c6884853b4/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudRunStepInfo.java#L63)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

Shoudn't this TerraformCloudRunType have a @JsonProperty?
</blockquote></details>

It's not necessary. When there isn't  @JsonProperty annotation, it looks for the variable name which is runType. Anyway, now it's added as we did it for others as well.
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudRunStepParameters.java` line 30 at r2](https://reviewable.io/reviews/harness/harness-core/43075#-NMnB4O--vwfc4GaxtjV:-NMoUl5I1e6L7LVrWexI:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/a7dfc5dfe0e1a8dd5ffd76648ef5c0c6884853b4/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/TerraformCloudRunStepParameters.java#L30)):*
<details><summary><i>Previously, tmacari-harness (Tudor Macari) wrote…</i></summary><blockquote>

let's get rid of stepFqn  we don't need to build any logic around it (we have a tech debt item to get rid of it in current terraform integration)
</blockquote></details>

done
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/runspecs/TerraformCloudRefreshSpec.java` line 49 at r2](https://reviewable.io/reviews/harness/harness-core/43075#-NMcy3UqEEtQMbLyS1Dp-r2-49:-NMoUmHr8-xFHSUUyQ4w:bphs38s) ([raw file](https://github.com/harness/harness-core/blob/a7dfc5dfe0e1a8dd5ffd76648ef5c0c6884853b4/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraformcloud/runspecs/TerraformCloudRefreshSpec.java#L49)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

Shouls we have @NotNull for discardPendingRuns? I saw in other classed you add this.
</blockquote></details>

it's not needed. When discardPendingRuns is not provided, we would assume that it's FALSE which is expected behavior

**Comment by alebuh** [comment]
trigger ft1

**Comment by alebuh** [comment]
trigger ft2

**Review by yogesh-chauhan — APPROVED** [review]

**Review by vlica-harness — APPROVED** [review]
> Reviewed 32 of 43 files at r1, 1 of 1 files at r2, 5 of 6 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43075)* status: 10 file(s) to review. 1 change requests. 6. unresolved discussions.

**Review by tmacari-harness — APPROVED** [review]
> Reviewed 5 of 6 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43075)* status: 3. unresolved discussions.

**Review by vlica-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43075)* status: :shipit: complete! ✓

**Pull request merged**